### PR TITLE
Finish mechanism for the update of ghost particles

### LINKED ISSUE #11093: Provide a mechanism to update ghost_particles location and properties without regenerating them
Dear all, 
I hope you are well. As always, thank you very much for taking the time.
Our recent benchmark on our usage of the particle methods has highlighted some limitations in the code when it runs in parallel. These could not have been planned or identified by @gassmoeller, there is absolutely not an ounce of blame in what follows. I think @gassmoeller work on particles is just awesome.

Let me describe our application case and things will be clearer. We are implementing a DEM method in Lethe using the particles structures (see : https://www.youtube.com/watch?v=__5UGx4fQps&feature=youtu.be for an example of DEM). On a single core, it has exactly the performance we need right now and everything is perfect (it about 10% faster than LIGGGHTS, a state of the art DEM code based on LAMMPS).

What is important to know about DEM is that we do not need to call sort_particles_into_subdomain_and_cells every iteration. We actually do it every X iterations, where X is the frequency at which we want to couple DEM and the background equations being solved on the mesh.

Currently, deal.II does not provide any mechanism to update the ghost particles location and properties without calling exchange_ghost_particles. However, exchange_ghosts is a very expensive operation. Here is the code for exchange_ghost_particles: https://www.dealii.org/developer/doxygen/deal.II/particle__handler_8cc_source.html#l01160

There are two sub issues to this issue.
The first one is that at every exchange_ghost_particles() the following structure is generated:

    
     std::vector<std::set<unsigned int>> vertex_to_neighbor_subdomain(
       triangulation->n_vertices());
     for (const auto &cell : triangulation->active_cell_iterators())
       {
         if (cell->is_ghost())
           for (const unsigned int v : GeometryInfo<dim>::vertex_indices())
             vertex_to_neighbor_subdomain[cell->vertex_index(v)].insert(
               cell->subdomain_id());
       }

This is a relatively expensive operation since we are doing it millions of times. This sub issue can be solved easily using a deal.II cache I believe, like my PR from last year on sort_particles_into_subdomain_and_cells(). No worries here.

The second issue is more profound. At every call, exchange_ghost_particles() clears the entire ghost_particles structure and rebuilds it from scratch, assuming that ghost particles may have changed cells and everything. The underlying assumption is that we are calling sort_particles_into_subdomain_and_cells at every time step. Regretfully, this is not what we are doing. 

My proposition would be to add a simple function called update_ghost_particles() which would only update the ghost particles location and property, without recreating the entire ghost particles data structure. Effectively it would "just" send the information from the owner of the real particles to the owner of the equivalent ghost particles and use this information to update the ghost location and property. However, this is significantly more challenging since this new function would need to know which processors own the ghost derived from real particles. I would like to do this cleanly and contribute it back to deal.II because I am sure it can be helpful to others down the line.

Would any of you have suggestions. tips or ideas as to what is the best way to implement this? I will do a PR for the first sub issue next week, but for the second part, I would greatly appreciate your advice (maybe @peterrum , @gassmoeller , @bangerth or @kronbichler  ?)

### PR DESCRIPTION
**Pull request by blaisb**
In this PR, we add a new feature that allow ghost particles to be recreated without recreating the map from scratch.
This is achieved by caching the necessary communication information in a new send_recv_cache_particles function (which at the moment has a lot of code duplication from send_recv_particles, but this was the least invasive way to do this).
Then we have a new function which only takes care of the update of particles themselves without altering the map.
The code could be made even more efficient by replacing the "destruction/deallocation/reconstruction" of the particles by a new Particle function that would "rewrite/update" the particle data. This would be done in  a follow-up PR.

The code is still a WIP, but it works. We have seen crazy performance improvement (from 2x to 6x as fast on 16 cores) on our applications. For us this PR is a game changer. Simulations that took days now take a few hours.

Right now the code is still in an ugly state (and I hate the fact that I had to copy over send_recv_particles). I will try to clean it ASAP, but I welcome any comments on how to make this better.
Again, this really makes me doubt that the multimap was the right structure for particle, but changing the particle container is out of the scope of the present PR. 

This closes #11093

### CODE REVIEW

File: include/deal.II/particles/particle_handler.h [hunk 1]
```diff
@@ -942,6 +942,72 @@ namespace Particles
...
+    struct GhostParticlesCache
+    {
+      bool                             valid = false;
+      std::vector<types::subdomain_id> neighbors;
+      std::vector<unsigned int>        send_pointers;
+      std::vector<unsigned int>        recv_pointers;
+      std::vector<typename std::multimap<internal::LevelInd,
+                                         Particle<dim, spacedim>>::iterator>
+        ghost_particles_iterators;
+    };
```
Comment (peterrum) [inline include/deal.II/particles/particle_handler.h:1007]: I am thinking about making this struct a bit more like the `Partitioner` class by moving the new methods here and renaming it to `ParticlePartitioner`.

File: source/particles/particle.cc [hunk 2]
```diff
@@ -180,6 +180,14 @@ namespace Particles
...
   }
 
+  template <int dim, int spacedim>
+  void
+  Particle<dim, spacedim>::free_properties()
+  {
+    if (property_pool != nullptr && properties != PropertyPool::invalid_handle)
+      property_pool->deallocate_properties_array(properties);
+  }
+
```
Comment (peterrum) [inline source/particles/particle.cc:190]: I am not that familiar with the particle infrastructure: why is this needed? What is the effect?

File: source/particles/particle_handler.cc [hunk 3]
```diff
@@ -1221,7 +1224,13 @@ namespace Particles
           }
       }
 
-    send_recv_particles(ghost_particles_by_domain, ghost_particles);
+    if (enable_cache)
+      {
+        send_recv_cache_particles(ghost_particles_by_domain, ghost_particles);
```
Comment (peterrum) [inline source/particles/particle_handler.cc:1229]: Maybe call it `send_recv_particles_and_build_cache`.

File: source/particles/particle_handler.cc [hunk 4]
```diff
@@ -1483,7 +1499,440 @@ namespace Particles
   }
 #endif
 
+#ifdef DEAL_II_WITH_MPI
+  template <int dim, int spacedim>
+  void
+  ParticleHandler<dim, spacedim>::send_recv_cache_particles(
```
Comment (peterrum) [inline source/particles/particle_handler.cc:1505]: I thinks we might be able to merge `send_recv_cache_particles` and `send_recv_cache`.  As far as I see it, you perform in `send_recv_cache_particles` and store some additional data if the construction of the cache is requested.

File: source/particles/particle_handler.cc [hunk 5]
```diff
@@ -1483,7 +1499,440 @@ namespace Particles
...
+    unsigned int particle_iterator_counter = 0;
+
+    while (reinterpret_cast<std::size_t>(recv_data_it) -
+             reinterpret_cast<std::size_t>(recv_data.data()) <
+           recv_pointers.back())
+      {
+        auto recv_particle =
+          ghost_particles_iterators[particle_iterator_counter];
+
+        recv_particle->second.free_properties();
```
Comment (peterrum) [inline source/particles/particle_handler.cc:1918]: I guess this is indeed needed?

File: include/deal.II/particles/particle_handler.h (hunk 1 above)
Comment (blaisb) [reply include/deal.II/particles/particle_handler.h:1007]: Makes sense, if it makes things more uniform :)

File: source/particles/particle.cc (hunk 2 above)
Comment (blaisb) [reply source/particles/particle.cc:190]: So this took me totally offguard. When you overwrite the particle to which an iterator points to, the destructor of that particle is not called. Consequently, it still has memory allocated in the memory pool and when the destructor of the memory pool is called, it crashes (because it asserts that the particles have de-allocated it's memory). Consequently, I had to add this routine to correctly clean the memory associated with particles before "overwriting them". Took me a while to figure this out. In the next PR I would replace this with a function that overwrites the data in the same memory space. This would prevent needless allocation/deallocation.

File: include/deal.II/base/mpi_tags.h [hunk 6]
```diff
@@ -109,6 +109,8 @@ namespace Utilities
           /// ParticleHandler<dim, spacedim>::send_recv_particles
           particle_handler_send_recv_particles_setup,
           /// ParticleHandler<dim, spacedim>::send_recv_particles
+          particle_handler_send_recv_particles_cache_setup,
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/deal.II/base/mpi_tags.h:112]: The comment is wrong. It would be a comment for send_recv_cache_particles. My bad

File: source/particles/particle_handler.cc (hunk 3 above)
Comment (blaisb) [reply source/particles/particle_handler.cc:1229]: Sure thing, makes sense!

File: source/particles/particle_handler.cc (hunk 4 above)
Comment (blaisb) [reply source/particles/particle_handler.cc:1505]: Yeah. We could definitely merge them. We could pass the boolean "build_cache" to send_receive_particles and it would take care of building the cache if needed. The added cost would be a  few extra ifs, but it would remove all code duplication. I just did not want to do this now because I was developing it and it was already a mess to understand what was happening. But definitely, now that it works and that I have the test, that could be a good idea.

File: source/particles/particle_handler.cc (hunk 5 above)
Comment (blaisb) [reply source/particles/particle_handler.cc:1918]: Yes! Otherwise the property_pool will crash when it's destructor is called because it still has allocated memory.

**Comment by blaisb** [comment]
@peterrum  If you have time (which I know is a rare thing) feel free to  make modifications directly in the original branch. I am absolutely open to any modifications as long as it keeps the feature there :).

**Comment by peterrum** [comment]
@blaisb I have made some simplifications to our code. In particular, now the cache is set up in `send_recv_particles` (if requested).

I was not able to push to this branch; instead I have opened a PR against this branch (see https://github.com/lethe-cfd/dealii/pull/5).

May I ask you take care of the documentation and the failing tests on GitHub (probably due to used variables...). Thx.

**Comment by blaisb** [comment]
> @blaisb I have made some simplifications to our code. In particular, now the cache is set up in `send_recv_particles` (if requested).
> 
> I was not able to push to this branch; instead I have opened a PR against this branch (see [lethe-cfd#5](https://github.com/lethe-cfd/dealii/pull/5)).
> 
> May I ask you take care of the documentation and the failing tests on GitHub (probably due to used variables...). Thx.

Absolutely. I'll take take of this early this week :)!
Thanks !

File: source/particles/particle_handler.cc [hunk 7]
```diff
@@ -1309,16 +1333,26 @@ namespace Particles
...
     // are send, because we might receive particles from other processes
     if (n_send_particles > 0)
       {
         // Allocate space for sending particle data
-        const unsigned int particle_size =
-          begin()->serialized_size_in_bytes() + cellid_size +
-          (size_callback ? size_callback() : 0);
-        send_data.resize(n_send_particles * particle_size);
+        send_data.resize(n_send_particles *
+                         individual_total_particle_data_size);
```
Comment (kronbichler) [inline source/particles/particle_handler.cc:1354]: Do we ever expect this number to be larger than 4G? `MPI` will not allow more than 2G or 4G per message as used below, so all I would suggest here is `Assert(static_cast<std::size_t>(n_send_particles) * individual_total_particle_data_size == static_cast<std::size_t>(n_send_particles * individual_total_particle_data_size), ExcMessage("Overflow when trying to send particle data));`

File: source/particles/particle_handler.cc [hunk 8]
```diff
@@ -1485,6 +1549,127 @@ namespace Particles
...
+    // Gather ghost particle iterators from the cache
+    auto &ghost_particles_iterators =
+      ghost_particles_cache.ghost_particles_iterators;
+    unsigned int particle_iterator_counter = 0;
+
+    while (reinterpret_cast<std::size_t>(recv_data_it) -
+             reinterpret_cast<std::size_t>(recv_data.data()) <
+           recv_pointers.back())
+      {
+        auto recv_particle =
```
Comment (kronbichler) [inline source/particles/particle_handler.cc:1651]: This is totally not your fault, but I wonder if these weird casts of pointers are the best approach. Maybe someone with more insight into the particles can explain me why this actually works?

**Review by kronbichler — COMMENTED** [review]
This looks very nice to me! I have two only tangentially related comments below.

**Comment by kronbichler** [comment]
> Right now the code is still in an ugly state (and I hate the fact that I had to copy over send_recv_particles). I will try to clean it ASAP, but I welcome any comments on how to make this better.

Regarding the code re-use, would it be an option to let the old function go into this function for the exchange part, after the communication has been constructed? At a superficial sight, it seems that one would not lose much by filling the communication pattern in a vector first, but I might be missing some details here.

**Comment by peterrum** [comment]
> Regarding the code re-use,

@kronbichler  The code has changed somewhat since the comment in the description of the PR. The function `send_recv_particles()` sets up the communication pattern (+communicates the particle IDs) and performs one round of exchange of properties and locations (all in one go); the function `send_recv_particles_properties_and_location()` uses the communication pattern set up by the other function. In my opinion, this is a nice compromise between the case that you need to exchange ghost particles during every time step (as originally designed for) and the case that you only want to update the properties every time step (as targeted by this PR). The code duplication is not that much as it used to be. @blaisb Did I interpret you comment correctly?

File: source/particles/particle_handler.cc (hunk 8 above)
Comment (blaisb) [reply source/particles/particle_handler.cc:1650]: I must be honest and say I have no clue how these work... :(

File: source/particles/particle_handler.cc (hunk 7 above)
Comment (blaisb) [reply source/particles/particle_handler.cc:1354]: This is such a great idea. I have never thought about this limitation.

**Comment by blaisb** [comment]
> > Regarding the code re-use,
> 
> @kronbichler The code has changed somewhat since the comment in the description of the PR. The function `send_recv_particles()` sets up the communication pattern (+communicates the particle IDs) and performs one round of exchange of properties and locations (all in one go); the function `send_recv_particles_properties_and_location()` uses the communication pattern set up by the other function. In my opinion, this is a nice compromise between the case that you need to exchange ghost particles during every time step (as originally designed for) and the case that you only want to update the properties every time step (as targeted by this PR). The code duplication is not that much as it used to be. @blaisb Did I interpret you comment correctly?

In the new  updated version by Peter and I code duplication has been completely eliminated :).
I am much more proud of this new version of it.

**Comment by blaisb** [comment]
@peterrum  and @kronbichler I fixed all the documentation and the tests. There might be bits there and there to clean, but I think this version is ready for review (and it works :) )!

File: source/particles/particle_handler.cc (hunk 8 above)
Comment (peterrum) [reply source/particles/particle_handler.cc:1657]: I think we should replace this weird structure and simply iterate over all entries in `ghost_particles_iterators`.

**Review by peterrum — APPROVED** [review]
Looks good to me. Do you want to add a change-log entry for you last few changes?

File: source/particles/particle_handler.cc [hunk 9]
```diff
@@ -1309,16 +1333,26 @@ namespace Particles
     std::vector<unsigned int> send_offsets(n_neighbors, 0);
     std::vector<char>         send_data;
 
+    // TODO: is size cachable?
+    const unsigned int individual_particle_data_size =
+      Utilities::MPI::max(n_send_particles > 0 ?
+                            ((begin()->serialized_size_in_bytes() +
+                              (size_callback ? size_callback() : 0))) :
+                            0,
+                          parallel_triangulation->get_communicator());
```
Comment (kronbichler) [inline source/particles/particle_handler.cc:1342]: I don't think the `MPI_Allreduce` behind `MPI::max` is all that bad compared to the other things that happen in this function. The only thing that would concern me is the synchronization point in case different MPI ranks run very far ahead others. That would suggest to skip the TODO. The other question is under which circumstances the `individual_particle_data_size` can actually change, which indeed would suggest to use caching.

File: source/particles/particle_handler.cc (hunk 7 above)
Comment (kronbichler) [reply source/particles/particle_handler.cc:1354]: To be on the safe side, which should then change the data type to `std::size_t` for `n_send_particles`. (I fear we will not easily be able to test whether the integer overflow is handled correctly because we would need to have huge amounts of data exchange.) But I think it is a good step to guard for the overflow.

**Review by kronbichler — APPROVED** [review]
Apart from a few minor comments, this looks good to me.

File: source/particles/particle_handler.cc (hunk 8 above)
Comment (kronbichler) [reply source/particles/particle_handler.cc:1657]: If that can be made to work, I would very much prefer it.

Comment (blaisb) [reply source/particles/particle_handler.cc:1657]: @gassmoeller  I would presume this originates from you? It would be nice to have your input here :)

File: source/particles/particle_handler.cc (hunk 9 above)
Comment (blaisb) [reply source/particles/particle_handler.cc:1342]: Those are good questions. The `individual_particle_data_size` can only change if the number of properties changes. I think it is not possible right now to change the number of properties without building a new particle_handler from scratch. I think right now this should not be a significant overhead since anyway it is followed by two sets of communications with a WaitAll so we are at a section where everything will need to synchronize anyway.

Comment (kronbichler) [reply source/particles/particle_handler.cc:1342]: Yes, let's keep things simple.

File: source/particles/particle_handler.cc (hunk 8 above)
Comment (blaisb) [reply source/particles/particle_handler.cc:1657]: For the loop in `send_recv_particles_positions_and_properties` I can modify it to use the vector to loop instead. For the regular `send_recv_particles` I don't have an alternative though :).

**Comment by blaisb** [comment]
@kronbichler  I think I addressed all of your suggestions (which were very good). I managed to replace the annoying loop over the recv_data by a simple loop over the ghost particles iterators. I also added a changelog. I supposed it was a major change and added Peter as co-author.

:) Thanks for the help with this! Once this is merged, I'll move onto trying to improve this a bit more. I think right now that recreating the particle from scratch is still a bit of an expensive operation, but I will need to profile more to make sure of that.

**Comment by blaisb** [comment]
The MPI build seems to have failed but I don't have access to the detail. It works on my machine (and it worked on OS X parallel) though... Is it a CI bug?

**Comment by tjhei** [comment]
> Is it a CI bug?

Maybe, I had to restart things and tests are running again.

**Review by peterrum — APPROVED** [review]
Would you mind to squash the commits?

**Comment by blaisb** [comment]
> Would you mind to squash the commits?

Done!
My evil plan to flood you with commits and to beat Wolfgang to top 1 contributor has been thwarted  ;)

**Review by peterrum — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]
